### DISCUSSION TITLE
Incorrect Input/Output values? kWh instead of Wh?

### DISCUSSION THREAD

**Discussion by AshleyJackson** [tolwi/hassio-ecoflow-cloud#87] [General]
Hello,

I own a Delta Pro and within the energy section of Home Assistant it is showing kWh generated and used.
![image](https://github.com/tolwi/hassio-ecoflow-cloud/assets/12793989/4a03305c-2c86-4ae2-969a-121de705b4de)

Is this value being converted incorrectly by the Integration?

Please let me know if you need any test data from me.
Thanks.

**Comment by knyghtryda** [comment]
@AshleyJackson @tolwi Just a reminder that first day integrations will always be strange, because the value stored is continually increasing.  That means the first time it gets picked up It will read your all time total.  After a day it will start to display your  delta value, not cumulative value.  Here are my day and week values:
<img width="2158" alt="Screenshot 2023-07-18 at 9 05 06 AM" src="https://github.com/tolwi/hassio-ecoflow-cloud/assets/1226173/b9809cfd-1aa6-4fc5-9233-2449a21b6efa">

<img width="2302" alt="Screenshot 2023-07-18 at 8 49 02 AM" src="https://github.com/tolwi/hassio-ecoflow-cloud/assets/1226173/cc6fdf82-fd0f-4f60-9c31-3c564f258941">

Note that due to some logging issues I had yesterday the daily solar in the second graph is kinda bugged out.  It looks like it grabbed the cumulative value again for some reason.

**Comment by tolwi** [comment]
hm, good point about initial accumulated values ...
can you please attach one more screnshot with `pd.chgSunPower` history for like 7 days

**Comment by knyghtryda** [comment]
Here's the last few days of Solar In Energy
<img width="2301" alt="Screenshot 2023-07-18 at 10 33 05 AM" src="https://github.com/tolwi/hassio-ecoflow-cloud/assets/1226173/23b849d2-389e-4e10-a3e9-92a5bb88c4f4">

And Solar Power
<img width="2299" alt="Screenshot 2023-07-18 at 10 54 42 AM" src="https://github.com/tolwi/hassio-ecoflow-cloud/assets/1226173/d8aaaf68-a1f1-45f2-baa8-4bdc93fbf023">

**Comment by knyghtryda** [comment]
@tolwi I figured out the energy spike issue, though I'm not exactly sure how to solve it.  What seems to be happening is that the Ecoflow will occasionally report a zero value for the energy.  It then goes back to reporting the normal energy value.  That just confuses the crap out of the Energy Dashboard which thinks it just went from a baseline of zero to full cumulative.  This probably means we need to either not report zero values at all, or take the previous value and use it if a value comes in at zero.  I'm not familiar enough with HA to know quite how to do either of those things though.

Here's the zero spikes
<img width="2298" alt="Screenshot 2023-07-18 at 2 23 06 PM" src="https://github.com/tolwi/hassio-ecoflow-cloud/assets/1226173/b5920da4-1e2f-4774-a1a7-03009ad2d88f">

Here's the energy spikes
<img width="2296" alt="Screenshot 2023-07-18 at 2 23 39 PM" src="https://github.com/tolwi/hassio-ecoflow-cloud/assets/1226173/764009ae-2637-4ba4-a887-293353ca5ef4">

**Reply by tolwi** [reply]
```
def _update_value(self, val: Any) -> bool:
        ival = int(val)
        if ival > 0:
          return super()._update_value(ival)
        else:
          return False
```
with 0 initial value this logic should be valid for all `SensorStateClass.TOTAL_INCREASING` sensors

**Reply by knyghtryda** [reply]
@tolwi This is where i put your code but I'm still seeing the value go wrong after a restart.
```
class EnergySensorEntity(BaseSensorEntity):
    _attr_device_class = SensorDeviceClass.ENERGY
    _attr_native_unit_of_measurement = UnitOfEnergy.WATT_HOUR
    _attr_state_class = SensorStateClass.TOTAL_INCREASING
    _attr_native_value = 0
    def _update_value(self, val: Any) -> bool:
        ival = int(val)
        if ival > 0:
          return super()._update_value(ival)
        else:
          return False
```

**Reply by tolwi** [reply]
try to remove `_attr_native_value = 0`

**Comment by AshleyJackson** [comment]
I've now gotten this twice today.
First time was when I restarted by Home Assistant server, second time was when I upgraded to v0.11.1.

![image](https://github.com/tolwi/hassio-ecoflow-cloud/assets/12793989/20270091-faf4-48b1-a71e-9779874ee10e)

**Reply by knyghtryda** [reply]
I've implemented @tolwi 's fix locally. I'll report here if it works, and then hopefully create a PR for it.

**Comment by vRzn** [comment]
i have a river pro, but it looks like there is no cumulative (Wh) value out of that device.
is it available for that device, or is it just select (larger) devices that have it?
i would love to add my river pro to my dashboard

**Reply by PaSch1** [reply]
You can use the helper "Riemann Summenintegralsensor" for this.
